### DISCUSSION TITLE
请问客户端多个vnext流量会如何路由？

### DISCUSSION THREAD

**Discussion by undefine-man** [XTLS/Xray-core#3070] [Q&A]
![屏幕截图 2024-02-28 051209](https://github.com/XTLS/Xray-core/assets/24623544/acaf8bcb-53a6-4d50-ad74-ed5b442e9da4)
是会将流量一分为多地发送给多个vnext吗？还是会只发送给第一个呢？

**Comment by vrnobody** [comment]
会随机抽一个vnext。以前v2ray 3.x没有balancer的时候曾经用过vnext来做负载均衡。

**Reply by undefine-man** [reply]
随机得到的vnext是跟session绑定的吗？这种随机的形式是不是会因为不同vnext节点的状况而产生有时候可以访问到某个网站，关闭网页标签后重新打开该网站却打不开这种情况

**Reply by vrnobody** [reply]
我不太确定你说的session具体指什么。

假设inbounds是socks协议，outbound是vmess协议并有3个vnext配置，那么core的代理过程是这样的：

浏览器访问一个网站会发起一条或者多条tcp链接，通过socks端口进入core。core经过一系列的if/else/routing判断，选出一个outbound。然后用类似math.rand(3)的算法，从vnext中随机选出一个服务器配置，接着拨号连接到相应的远程服务器。这时候inbound拥有一条来自浏览器的tcp链接，outbound拥有一条到远程服务器的链接，最后启动一个线程/协程在两条链接之间拷贝数据。这是一个极度简化的过程，忽略了身份验证、加密、udp数据处理等等。

简单来说core是给传入的tcp链接随机选vnext，和其他状态无关。

**Reply by undefine-man** [reply]
根据描述，我在源码的app/router文件夹下找到了strategy_random.go文件，其中的PickOutbound与你的描述相符。
![image](https://github.com/XTLS/Xray-core/assets/24623544/638d80df-e8fc-4fb8-b7d6-83017ea6e1e2)


可是我通过断点和单步调试的手段，发现在两个vnext的情况（vnext节点均没有问题）下，代码并没有走到app/router的PickOutbound这个函数。
与outbound、vnext相关的，只发现走了common/protocol/server_picker.go下的PickServer函数（此PickServer隶属于RoundRobinServerPicker接口
![屏幕截图 2024-03-02 000535](https://github.com/XTLS/Xray-core/assets/24623544/0cde8730-be35-4899-8fb8-e1b0b723156c)
根据隶属的接口名RoundRobinServerPicker及PickServer函数的行为，似乎采用的是一种轮询的形式？


后通过断点inbound进来的流量信息，单个页面访问www.google.com并追踪，一个inbound流量会通过两次PickServer函数。
我不太确定我的操作，但根据现象，单个inbound好像会分裂为两个outbound流量？并且没有进入router。
可以请求操作验证一下吗

**Reply by vrnobody** [reply]
看了一下代码，你的调试结果是对的，我搞错了。挑选vnext采用轮询算法而不是随机抽取。至于一个inbound为什么分裂成两个的问题我也没仔细研究过，不太清楚。

**Comment by xqzr** [comment]
轮询

**Comment by us254** [comment]
it appears that V2Ray supports both random and round-robin load balancing strategies for selecting outbound VNext nodes.

The strategy_random.go file contains the implementation of the RandomStrategy struct, which has a PickOutbound method that randomly selects an outbound VNext node from the list of candidates using the dice.Roll function.

On the other hand, the server_picker.go file contains the implementation of the RoundRobinServerPicker struct, which implements a round-robin algorithm for selecting outbound VNext nodes. The PickServer method in this struct cycles through the list of available servers in a fixed order, returning the next server in the list for each call.

The configuration file you provided seems to be using the vless protocol with multiple VNext nodes specified in the vnext array. However, it doesn't explicitly specify the load balancing strategy to be used.

If the load balancing strategy is not explicitly specified in the configuration, V2Ray likely uses a default strategy. Based on your observation that the PickServer function from the RoundRobinServerPicker is being called, it appears that the default load balancing strategy in your case is round-robin.

To change the load balancing strategy to random, you would need to modify the configuration file and add the balancingStrategy option under the outbounds section, like this:
```
"outbounds": [
  {
    "protocol": "vless",
    "settings": {
      "vnext": [
        // VNext nodes
      ]
    },
    "balancingStrategy": "random"
  }
]
```

**Reply by undefine-man** [reply]
Thanks for reply.
I wonder how did you know there is a "balancingStrategy" field supported in the config file?
I couldn't see anything of this field on Xray project doc website. Are you acknowledge it from source code? or somewhere else?
I am trying to implement a new function on Xray project after completely understand how xray project works , but it seems document is not enough details.
